jgrapht-core/src/main/java/org/jgrapht/alg/decomposition/HeavyPathDecomposition.java

```diff
@@ -0,0 +1,511 @@
+
+        Arrays.fill(parent, -1);
+        Arrays.fill(path, -1);
+        Arrays.fill(depth, -1);
+        Arrays.fill(component, -1);
+        Arrays.fill(positionInPath, -1);
+
+        /*
+            Iterate through all roots and compute the paths for each tree individually
+         */
+        int numberComponent = 0;
+        for (V root: roots){
+            Integer u = vertexMap.get(root);
+
+            if (u == null)
+                throw new IllegalArgumentException("root: " + root + " not contained in graph");
+
+            if (component[u] == -1) {
+                dfsIterative(u, numberComponent++);
+            }
+        }
+
+        firstNodeInPath = new int[numberOfPaths];
+
+        /*
+            Reverse the position of all vertices that are present in some path.
+            After this the positionInPath[u] = 0 if u is the first node in the path (i.e. the node closest to the root)
+
+            Also compute firstNodeInPath[i] = u such that path[u] = i and positionInPath[u] = 0
+         */
+        for (int i = 0; i < graph.vertexSet().size(); i++){
+            if (path[i] != -1){
+                positionInPath[i] = lengthPath[path[i]] - positionInPath[i] - 1;
+
+                if (positionInPath[i] == 0)
+                    firstNodeInPath[path[i]] = i;
+            }
+        }
+
+        /*
+            Compute the paths as unmodifiable data structures (list)
+         */
+        List<List<V>> paths = new ArrayList<>(numberOfPaths);
+
+        for (int i = 0; i < numberOfPaths; i++) {
+            List<V> path = new ArrayList<>(lengthPath[i]);
+
+            for (int j = 0; j < lengthPath[i]; j++) {
+                path.add(null);
+            }
+
+            paths.add(path);
+        }
+
+        for (int i = 0; i < graph.vertexSet().size(); i++) {
+            if (path[i] != -1){
+                paths.get(path[i]).set(positionInPath[i], indexList.get(i));
+            }
+        }
+
+        for (int i = 0; i < numberOfPaths; i++) {
+            paths.set(i, Collections.unmodifiableList(paths.get(i)));
+        }
+
+        this.paths = Collections.unmodifiableList(paths);
+    }
+
+    /**
+     * Set of heavy edges.
+     *
+     * @return (immutable) set of heavy edges
+     */
+    public Set<E> getHeavyEdges(){
+        return Collections.unmodifiableSet(this.heavyEdges);
+    }
+
+    /**
+     * Set of light edges.
+     *
+     * @return (immutable) set of light edges
+     */
+    public Set<E> getLightEdges(){
+        return Collections.unmodifiableSet(
+                graph.edgeSet().stream().filter(n -> !this.heavyEdges.contains(n)).collect(Collectors.toSet()));
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public PathDecomposition<V, E> getPathDecomposition() {
+        return new PathDecompositionImpl<>(graph, getHeavyEdges(), this.paths);
+    }
+
+    /**
+     * Return the internal representation of the data.
+     *
+     * Note: For developer use only
```

Comment:
Add one more comment here which states that we are returning direct references to the internal data structures and that this is done only for efficiency reasons.